### PR TITLE
5.7 - Fixed PS-7346 - Huge memory usage for audit plugin when large queries…

### PR DESCRIPTION
**Pull request by altmannmarcelo**
… are executed

https://jira.percona.com/browse/PS-7346

Function audit_log_general_record can be called twice for each record.
1st time it is called with a 4096k buffer. It will estimate via
query_length what can be the maximum size of incoming query if every
character has to be converted to utf8mb4(basically 4 times its size).
If this size fits the buffer, it will convert event->general_query.str
to utf8mb4, use the current buffer to store the converted query and
advance the return pointer by query_length bytes. After we have the real
query, we escape it.
Otherwise (if the buffer is not big enough to convert the query), we
will fall into the else branch. Here we calculate the size of the
current query after escaping it and multiply by 4 (since we have not
converted the query to utf8mb4) + the maximum size of incoming query if
every character has to be converted to utf8mb4.
We continue by escaping user / host / external_user / ip and db storing
the required bytes into buflen_estimated.
After this point we do a check to validate if the current buffer is big
enough to hold buflen_estimated (the converted query + the full output
of audit log entry). In case we don't have a big enough buffer, we set
outlen to buflen_estimated and return NULL. Then audit_log_notify will
allocate outlen bytes into its buffer and call audit_log_general_record
for the 2nd time, this time with the resized buffer.

Issue:
Before 9fd57d3 we estimated buflen_estimated as twice of the output. If
it was not enough we would increase the size of buffer by 4 times before
calling audit_log_general_record the second time.
After 9fd57d3 the else branch has been enhance to make a better estimate
of the size of the output + size of the query, taking into consideration
utf8mb4, however it missed two points:

buflen_estimated was doubling the size. This is no longer required
since the else branch already take into account escaped utf8mb4 query +
non escaped utf8mb4 query.
When estimating the size of the non escaped utf8mb4 query, it was
wrongly multiplying its size by my_charset_utf8mb4_general_ci.mbmaxlen,
however, this computation has already been performed before the if/else.

### CODE REVIEW

**Comment by altmannmarcelo** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-param/343/ - full run
https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/25/ - ASAN
https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/26/ - Valgrind

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# 10M query
+let $very_long_text=`SELECT REPEAT('a', 10485760)`;
+eval INSERT INTO test (col1) VALUES ('$very_long_text');
+--enable_query_log
+--enable_result_log
+--let $assert_text = Memory usage is less than 100M
+--let $query_result = query_get_value(SELECT current_alloc FROM sys.memory_global_by_current_bytes WHERE event_name = 'memory/sql/THD::variables', current_alloc, 1)
+--let $assert_cond= "$query_result" <= 100
+--source include/assert.inc
+DROP TABLE test;
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test:23]: @altmannmarcelo  Should there be some kind of disabling performance schema instrumentation  after you did
```
UPDATE performance_schema.setup_instruments
SET ENABLED = 'YES'
WHERE NAME LIKE 'memory/%';
```
`mtr-check` may not detect this but it's possible that other tests running on the same mysqld instance will be affected.

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 2]
```diff
@@ -0,0 +1,25 @@
+--source include/not_embedded.inc
+--source include/have_perfschema.inc
+
+let $MYSQLD_DATADIR= `select @@datadir`;
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test]:
```suggestion
--let $MYSQLD_DATADIR= `SELECT @@datadir`
```

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+let $log_file=$MYSQLD_DATADIR/test_audit.log;
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test]:
```suggestion
--let $log_file=$MYSQLD_DATADIR/test_audit.log
```

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+--remove_file $log_file
+SET GLOBAL audit_log_flush=ON;
+
+--disable_query_log
+--disable_result_log
+UPDATE performance_schema.setup_instruments
+SET ENABLED = 'YES'
+WHERE NAME LIKE 'memory/%';
+CREATE TABLE test (col1 longblob NOT NULL);
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test]:
```suggestion
CREATE TABLE test (col1 LONGBLOB NOT NULL);
```

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# 10M query
+let $very_long_text=`SELECT REPEAT('a', 10485760)`;
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test]:
```suggestion
--let $very_long_text=`SELECT REPEAT('a', 10485760)`
```

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+eval INSERT INTO test (col1) VALUES ('$very_long_text');
+--enable_query_log
+--enable_result_log
+--let $assert_text = Memory usage is less than 100M
+--let $query_result = query_get_value(SELECT current_alloc FROM sys.memory_global_by_current_bytes WHERE event_name = 'memory/sql/THD::variables', current_alloc, 1)
+--let $assert_cond= "$query_result" <= 100
```
Comment (percona-ysorokin) [inline plugin/audit_log/tests/mtr/audit_log_memory_consumption.test]:
```suggestion
--let $assert_cond= "[SELECT current_alloc FROM sys.memory_global_by_current_bytes WHERE event_name = \'memory/sql/THD::variables\']" <= 100
```

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]
One comment regarding disabling memory instrumentation
and a few regarding consistency in `--command` vs `command;` style;

File: plugin/audit_log/tests/mtr/audit_log_memory_consumption.test (hunk 1 above)
Comment (altmannmarcelo) [reply plugin/audit_log/tests/mtr/audit_log_memory_consumption.test:23]: @percona-ysorokin I've checked other tests that change setup_instruments and the clean-up is enable everything:
few examples:
https://github.com/percona/percona-server/blob/8.0/mysql-test/suite/sysschema/include/ps_setup_instruments_cleanup.inc
https://github.com/percona/percona-server/blob/8.0/mysql-test/suite/perfschema/include/memory_aggregate_cleanup.inc#L40

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**